src/puppetlabs/puppetdb/queue.clj

```diff
@@ -257,18 +301,39 @@
       (throw (IllegalArgumentException. (trs "Cannot enqueue item of type {0}" (class item)))))
 
     (let [^CommandRef cmdref item
-          command-type (:command cmdref)
-          certname (:certname cmdref)]
+          {command-type :command
+           certname :certname
+           producer-ts :producer-ts} cmdref]
 
-      (when (or (= command-type "replace catalog")
-                (= command-type "replace facts"))
-        (when-let [^CommandRef old-command (.get certnames-map [command-type certname])]
-          (.put fifo-queue
-                (:id old-command)
-                (assoc old-command :delete? true))
-          (delete-update-fn (:command old-command) (:version old-command)))
-        (.put certnames-map [command-type certname] cmdref))
-      (.put fifo-queue (:id cmdref) cmdref))
+      (if (and
+           producer-ts
+           (or (= command-type "replace catalog")
+               (= command-type "replace facts")))
+        (let [^CommandRef maybe-old-command (.get certnames-map [command-type certname])]
+          (cond
+            (nil? maybe-old-command)
+            (do
+              (.put certnames-map [command-type certname] cmdref)
+              (.put fifo-queue (:id cmdref) cmdref))
+
+            (time/after? producer-ts
+                         (:producer-ts maybe-old-command))
+            (do
+              (.put certnames-map [command-type certname] cmdref)
+              (.put fifo-queue
+                    (:id maybe-old-command)
+                    (assoc maybe-old-command :delete? true))
+              (.put fifo-queue (:id cmdref) cmdref)
+              (delete-update-fn (:command maybe-old-command) (:version maybe-old-command)))
+
+            :else
+            (do
+              (.put fifo-queue
+                    (:id cmdref)
+                    (assoc cmdref :delete? true))
```

Comment:
I remembered why, just as I was going in to change this. This part of the code is putting a Command Ref in the channel, which means the command is already on disk. To avoid storing this would be pretty difficult with the current code. We'd need to expose the data outside of the mutex in a way that also locks it, consult the map and then decide to discard.